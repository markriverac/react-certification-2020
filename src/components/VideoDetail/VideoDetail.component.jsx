import React, { useContext } from 'react';
import './VideoDetail.styles.css';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import FavoriteButton from '../FavoriteButton';
import UserContext from '../../providers/UserProvider/User.context';
import VideoContext from '../../providers/VideoProvider/Video.context';

function VideoDetail() {
  const { userState } = useContext(UserContext);
  const { videoState } = useContext(VideoContext);

  if (!videoState.selectedVideo.id) {
    return <></>;
  }

  const video = videoState.selectedVideo;

  const videoSrc = `https://www.youtube.com/embed/${video.id.videoId}`;
  return (
    <div className="video-container">
      <div className="frame-container">
        <iframe className="video" src={videoSrc} allowFullScreen title="Video player" />
      </div>
      <Row className="header">
        <Col>
          <h4> {video.snippet.title}</h4>
        </Col>
        <Col xs="auto">
          {userState.isAuth && (
            <FavoriteButton video={video} variant="outline-warning">
              Add to favorite
            </FavoriteButton>
          )}
        </Col>
      </Row>
      <p> {video.snippet.description}</p>
    </div>
  );
}

export default VideoDetail;
