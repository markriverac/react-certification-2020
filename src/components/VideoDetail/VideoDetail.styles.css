.frame-container {
  text-align: center;
}

.video {
  width: 60vw;
  height: 60vh;
}

.video-description {
  display: -webkit-box;
  margin: auto;
  overflow: hidden;
  width: 100%;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;  
}
