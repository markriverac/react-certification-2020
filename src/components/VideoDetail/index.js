/* istanbul ignore file */
export { default } from './VideoDetail.component';
