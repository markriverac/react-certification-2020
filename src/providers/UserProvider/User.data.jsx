/* istanbul ignore file */
export const userData = { isAuth: false, name: '', favorites: [] };
