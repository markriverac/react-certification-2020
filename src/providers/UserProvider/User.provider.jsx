import React, { useState } from 'react';
import UserContext from './User.context';
import { userData } from './User.data';

function UserProvider({ children }) {
  const [userState, setuserState] = useState(userData);

  function logIn(username) {
    setuserState({ ...userState, name: username, isAuth: true });
  }
  function logOut() {
    setuserState({ ...userState, isAuth: false });
  }
  function addFavorite(video) {
    const newFavorites = userState.favorites;
    newFavorites.push(video);
    setuserState({ ...userState, favorites: newFavorites });
  }
  function removeFavorite(video) {
    const favoriteVideos = userState.favorites;
    const videoIndex = favoriteVideos.indexOf(video);
    if (videoIndex > -1) {
      favoriteVideos.splice(videoIndex, 1);
      setuserState({ ...userState, favorites: favoriteVideos });
    }
  }

  return (
    <UserContext.Provider
      value={{ userState, logIn, logOut, addFavorite, removeFavorite }}
    >
      {children}
    </UserContext.Provider>
  );
}

export default UserProvider;
