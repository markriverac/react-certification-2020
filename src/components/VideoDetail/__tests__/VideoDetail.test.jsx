import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import UserContext from '../../../providers/UserProvider/User.context';
import VideoDetail from '../VideoDetail.component';
import UserProvider from '../../../providers/UserProvider/User.provider';

Enzyme.configure({ adapter: new Adapter() });
const videoDetail = {
  id: 'id',
  snippet: {
    localized: { title: 'title' },
    description: 'description',
  },
};

it('Renders without exploding', () => {
  const videoDetailComponent = renderer.create(
    <UserProvider>
      <VideoDetail videoDetail={videoDetail} />
    </UserProvider>
  );
  const tree = videoDetailComponent.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders favorite button when logged in', () => {
  const userState = { isAuth: true, favorites: [] };
  const videoDetailComponent = renderer.create(
    <UserContext.Provider value={{ userState }}>
      <VideoDetail videoDetail={videoDetail} />
    </UserContext.Provider>
  );
  const tree = videoDetailComponent.toJSON();
  expect(tree.children[1].children[1].children[0].type).toEqual('button');
});

it('does not render favorite button when not logged in', () => {
  const userState = { isAuth: false };
  const videoDetailComponent = renderer.create(
    <UserContext.Provider value={{ userState }}>
      <VideoDetail videoDetail={videoDetail} />
    </UserContext.Provider>
  );
  const tree = videoDetailComponent.toJSON();
  expect(tree.children[1].children[1].children).toEqual(null);
});

it('Renders without exploding when no videodetailid', () => {
  const videoNull = videoDetail;
  videoNull.id = null;
  const videoDetailComponent = renderer.create(
    <UserProvider>
      <VideoDetail videoDetail={videoNull} />
    </UserProvider>
  );
  const tree = videoDetailComponent.toJSON();
  expect(tree).toMatchSnapshot();
});
