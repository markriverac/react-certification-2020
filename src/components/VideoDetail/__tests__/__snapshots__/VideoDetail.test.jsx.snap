// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without exploding 1`] = `
<div
  className="video-container"
>
  <div
    className="frame-container"
  >
    <iframe
      allowFullScreen={true}
      className="video"
      src="https://www.youtube.com/embed/id"
      title="Video player"
    />
  </div>
  <div
    className="header row"
  >
    <div
      className="col"
    >
      <h4>
         
        title
      </h4>
    </div>
    <div
      className="col-auto"
    />
  </div>
  <p
    className="video-description"
  >
     
    description
  </p>
</div>
`;

exports[`Renders without exploding when no videodetailid 1`] = `null`;
